### PR TITLE
Add LZ4 compressor with Java9 perf improvements

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Java9 added a number of features that are useful to improve compression
and decompression. These include the Arrays#mismatch method and
VarHandles. This commit adds compression tools forked from the java-lz4
library which include these improvements. We hope to contribute these
changes back to the original project, however the project currently
supports Java7 so this is not possible at the moment.

### CODE REVIEW

**Comment by tbrooks8** [comment]
This adds support on 7.x by using a multi-release jar. @rjernst / @mark-vieira - hopefully someone can confirm if I structured the grade.build file properly.

**Review by rjernst — APPROVED** [review]
The multi release jar changes look fine to me

**Comment by mark-vieira** [comment]
I'm really not a fan of mult-release jars. I think simply instantiating the correct implementation at runtime via reflection is much more obvious and simpler. I'm also not sure which scenario is going to be more of a pain in terms of backporting, the MR scenario where our custom implementation now lives in different places in different branches or the reflection one, where `Compression` will different between branches. I'm leaning towards the latter, that building this custom fork as a MR jar in one branch but not in the other is going to be a bigger overall pain.

I think my preference would be to simply build `elasticsearch-lz4` lib as a normal jar, like we do in `master` and just set the target compatibility to Java 9. Then in `:server` when we go to initialize the compression implementation, we check the current Java version and create the appropriate instance via reflection.

File: libs/lz4/src/test/java/org/elasticsearch/lz4/ESLZ4CompressorTests.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    public void testCompressRealisticUnicode() {
+        for (int i = 0; i < 15; ++i) {
+            int stringLengthMultiplier = randomFrom(5, 10, 20, 40, 80, 160, 320);
+
+            final String uncompressedString = randomRealisticUnicodeOfCodepointLength(stringLengthMultiplier * 1024);
+            byte[] uncompressed = uncompressedString.getBytes(StandardCharsets.UTF_8);
+
+            byte[] compressed = new byte[uncompressed.length + uncompressed.length / 255 + 16];
+            byte[] unForkedCompressed = new byte[uncompressed.length + uncompressed.length / 255 + 16];
+            LZ4Compressor compressor = ESLZ4Compressor.INSTANCE;
```
Comment (henningandersen) [inline libs/lz4/src/test/java/org/elasticsearch/lz4/ESLZ4CompressorTests.java:34]: I wonder if we can validate that on java 9+, we get our own variant?

**Review by henningandersen — APPROVED** [review]
LGTM.

File: libs/lz4/src/test/java/org/elasticsearch/lz4/ESLZ4CompressorTests.java (hunk 1 above)
Comment (mark-vieira) [reply libs/lz4/src/test/java/org/elasticsearch/lz4/ESLZ4CompressorTests.java:34]: We already run tests with various runtimes. Perhaps we could add a unit test with an `assume` condition based on Java runtime version that verifies the type returned here?

**Review by breskeby — CHANGES_REQUESTED** [review]
I really really think we should avoid Multi-release JARs in our project. They address a real problem but I strongly think they are the wrong solution to this problem. Correct modeling of dependencies, as well as coupling of artifacts and variants as mentioned before should be preferred. I suggest we take @mark-vieira approach he described in his comment. That would also make testing easier and is overall the simplier solution

**Review by breskeby — APPROVED** [review]
Minor polishing request about the build.gradle file. otherwise looks good to me.

**Pull request merged**